File: NextApp/iOS/OrgPortalNext/Core/Notifications/NotificationService.swift
import Combine
import CryptoKit
import Foundation
import Model
import UIKit
import UserNotifications

@MainActor
public protocol ScheduleNotificationScheduling: AnyObject {
    func synchronizeReminder(for schedule: Schedule) async throws
    func removeReminder(scheduleId: UUID) async
}

@MainActor
public final class UserNotificationScheduler: ScheduleNotificationScheduling {
    private let center: UNUserNotificationCenter

    public init(center: UNUserNotificationCenter = .current()) {
        self.center = center
    }

    public func synchronizeReminder(for schedule: Schedule) async throws {
        let identifier = Self.identifier(for: schedule.id)
        center.removePendingNotificationRequests(withIdentifiers: [identifier])

        guard let reminder = schedule.reminderSetting, reminder.isEnabled else {
            return
        }
        let granted = try await center.requestAuthorization(options: [.alert, .sound])
        guard granted else { return }
        let fireDate = schedule.startDateTime.addingTimeInterval(
            TimeInterval(-reminder.notifyBeforeMinutes * 60)
        )
        guard fireDate > .now else { return }

        let content = UNMutableNotificationContent()
        content.title = schedule.title
        content.body = schedule.location
        content.sound = .default
        content.userInfo = ["scheduleId": schedule.id.uuidString]
        let components = Calendar.current.dateComponents(
            [.year, .month, .day, .hour, .minute],
            from: fireDate
        )
        let trigger = UNCalendarNotificationTrigger(
            dateMatching: components,
            repeats: false
        )
        try await center.add(
            UNNotificationRequest(
                identifier: identifier,
                content: content,
                trigger: trigger
            )
        )
    }

    public func removeReminder(scheduleId: UUID) async {
        center.removePendingNotificationRequests(
            withIdentifiers: [Self.identifier(for: scheduleId)]
        )
    }

    private static func identifier(for id: UUID) -> String {
        "schedule-reminder-\(id.uuidString)"
    }
}

public struct FcmToken: Identifiable, Equatable, Codable, Sendable {
    public enum OperatingSystem: String, Codable, Sendable {
        case iOS
        case android = "Android"
    }

    public enum Environment: String, Codable, Sendable {
        case development
        case production
    }

    public var id: String
    public var userId: String
    public var token: String
    public var appVariant: String
    public var os: OperatingSystem
    public var environment: Environment
    public var updatedAt: Date

    public init(
        id: String,
        userId: String,
        token: String,
        appVariant: String = "next",
        os: OperatingSystem,
        environment: Environment,
        updatedAt: Date = .now
    ) {
        self.id = id
        self.userId = userId
        self.token = token
        self.appVariant = appVariant
        self.os = os
        self.environment = environment
        self.updatedAt = updatedAt
    }

    public static func id(for token: String) -> String {
        SHA256.hash(data: Data(token.utf8)).map { String(format: "%02x", $0) }.joined()
    }
}

public enum NotificationType: String, Codable, Sendable, CaseIterable {
    case announcement
    case adminReply = "admin_reply"
    case videoQuestionAnswer = "video_question_answer"
    case event
    case supportMessage = "support_message"

    fileprivate var host: String {
        switch self {
        case .announcement: "announcements"
        case .adminReply: "posts"
        case .videoQuestionAnswer: "video-questions"
        case .event: "events"
        case .supportMessage: "support"
        }
    }

    fileprivate var targetKey: String {
        switch self {
        case .announcement: "announcementId"
        case .adminReply: "postId"
        case .videoQuestionAnswer: "questionId"
        case .event: "eventId"
        case .supportMessage: "supportMessageId"
        }
    }
}

public struct NotificationRoute: Equatable, Codable, Sendable {
    public let type: NotificationType
    public let targetId: String
    public let communityId: String?

    public init(type: NotificationType, targetId: String, communityId: String? = nil) {
        self.type = type
        self.targetId = targetId
        self.communityId = communityId
    }
}

public struct NotificationNavigationDecision: Equatable, Sendable {
    public let route: NotificationRoute
    public let communityIdToSelect: String?

    public static func resolve(
        route: NotificationRoute,
        selectedCommunityId: String?
    ) -> Self {
        Self(
            route: route,
            communityIdToSelect: route.communityId == selectedCommunityId
                ? nil
                : route.communityId
        )
    }
}

public enum NotificationRouteParser {
    public static let typeKey = "notificationType"
    public static let deepLinkKey = "deepLink"
    public static let communityIdKey = "communityId"

    public static func url(for route: NotificationRoute) -> URL {
        var components = URLComponents()
        components.scheme = "orgportalnext"
        components.host = route.type.host
        let encodedTarget = route.targetId.addingPercentEncoding(
            withAllowedCharacters: .urlPathAllowed
        ) ?? route.targetId
        components.percentEncodedPath = route.type == .supportMessage
            ? "/messages/\(encodedTarget)"
            : "/\(encodedTarget)"
        if let communityId = route.communityId {
            components.queryItems = [URLQueryItem(name: communityIdKey, value: communityId)]
        }
        return components.url!
    }

    public static func route(userInfo: [AnyHashable: Any]) -> NotificationRoute? {
        if let value = userInfo[deepLinkKey] as? String,
           let url = URL(string: value),
           let route = route(url: url) {
            return route
        }
        guard let rawType = userInfo[typeKey] as? String,
              let type = NotificationType(rawValue: rawType) else { return nil }
        let targetId = (userInfo[type.targetKey] as? String)
            ?? (userInfo["targetId"] as? String)
            ?? (type == .supportMessage ? userInfo["messageId"] as? String : nil)
        guard let targetId, !targetId.isEmpty else { return nil }
        return NotificationRoute(
            type: type,
            targetId: targetId,
            communityId: (userInfo[communityIdKey] as? String)?.nilIfEmpty
        )
    }

    public static func route(url: URL) -> NotificationRoute? {
        guard url.scheme == "orgportalnext",
              let type = NotificationType.allCases.first(where: { $0.host == url.host }) else {
            return nil
        }
        let segments = url.pathComponents.filter { $0 != "/" }
        let targetId: String?
        if type == .supportMessage {
            targetId = segments.count == 2 && segments[0] == "messages" ? segments[1] : nil
        } else {
            targetId = segments.count == 1 ? segments[0] : nil
        }
        guard let targetId, !targetId.isEmpty else { return nil }
        let communityId = URLComponents(url: url, resolvingAgainstBaseURL: false)?
            .queryItems?.first(where: { $0.name == communityIdKey })?.value?.nilIfEmpty
        return NotificationRoute(type: type, targetId: targetId, communityId: communityId)
    }
}

@MainActor
public final class NotificationRouter: NSObject, ObservableObject,
    UNUserNotificationCenterDelegate {
    public nonisolated static let messageIdKey = "supportMessageId"
    public nonisolated static let typeKey = NotificationRouteParser.typeKey
    public nonisolated static let supportType = NotificationType.supportMessage.rawValue
    @Published public private(set) var route: NotificationRoute?
    @Published public private(set) var routeSequence = 0
    private let center: UNUserNotificationCenter

    public var messageId: String? {
        route?.type == .supportMessage ? route?.targetId : nil
    }

    public init(center: UNUserNotificationCenter = .current()) {
        self.center = center
        super.init()
        center.delegate = self
        Task {
            let granted = (try? await center.requestAuthorization(
                options: [.alert, .sound, .badge]
            )) ?? false
            if granted { UIApplication.shared.registerForRemoteNotifications() }
        }
    }

    public func open(url: URL) {
        publish(NotificationRouteParser.route(url: url))
    }

    public nonisolated static func messageId(userInfo: [AnyHashable: Any]) -> String? {
        if let value = userInfo[messageIdKey] as? String, !value.isEmpty {
            return value
        }
        return NotificationRouteParser.route(userInfo: userInfo)
            .flatMap { $0.type == .supportMessage ? $0.targetId : nil }
    }

    public nonisolated static func messageId(url: URL) -> String? {
        NotificationRouteParser.route(url: url)
            .flatMap { $0.type == .supportMessage ? $0.targetId : nil }
    }

    nonisolated public func userNotificationCenter(
        _ center: UNUserNotificationCenter,
        didReceive response: UNNotificationResponse
    ) async {
        let route = NotificationRouteParser.route(
            userInfo: response.notification.request.content.userInfo
        )
        await MainActor.run { self.publish(route) }
    }

    nonisolated public func userNotificationCenter(
        _ center: UNUserNotificationCenter,
        willPresent notification: UNNotification
    ) async -> UNNotificationPresentationOptions {
        [.banner, .sound, .badge]
    }

    private func publish(_ route: NotificationRoute?) {
        guard let route else { return }
        self.route = route
        routeSequence += 1
    }
}

public typealias SupportNotificationRouter = NotificationRouter

public enum SupportNotificationContent {
    public static func make(messageId: String) -> UNMutableNotificationContent {
        let content = UNMutableNotificationContent()
        content.title = "サポート"
        content.body = "サポートから新しいメッセージがあります"
        content.sound = .default
        content.userInfo = [
            NotificationRouter.messageIdKey: messageId,
            NotificationRouter.typeKey: NotificationRouter.supportType
        ]
        return content
    }
}

public protocol FcmTokenStoring: Sendable {
    func save(_ token: FcmToken, idToken: String) async throws
    func delete(userId: String, tokenId: String, idToken: String) async throws
}

public protocol FcmRegistrationTokenProviding: Sendable {
    func currentToken() async throws -> String
}

public actor FcmTokenRegistrationManager {
    private struct Registration: Sendable {
        let userId: String
        let idToken: String
        let tokenId: String
    }

    private let store: any FcmTokenStoring
    private let tokenProvider: any FcmRegistrationTokenProviding
    private let environment: FcmToken.Environment
    private let now: @Sendable () -> Date
    private var registration: Registration?

    public init(
        store: any FcmTokenStoring,
        tokenProvider: any FcmRegistrationTokenProviding,
        environment: FcmToken.Environment,
        now: @escaping @Sendable () -> Date = { Date() }
    ) {
        self.store = store
        self.tokenProvider = tokenProvider
        self.environment = environment
        self.now = now
    }

    public func synchronize(userId: String?, idToken: String?) async throws {
        guard let userId, !userId.isEmpty,
              let idToken, !idToken.isEmpty else {
            if let registration {
                try await store.delete(
                    userId: registration.userId,
                    tokenId: registration.tokenId,
                    idToken: registration.idToken
                )
            }
            registration = nil
            return
        }
        if let previous = registration, previous.userId != userId {
            try await store.delete(
                userId: previous.userId,
                tokenId: previous.tokenId,
                idToken: previous.idToken
            )
            registration = nil
        }
        let token = try await tokenProvider.currentToken()
        try await register(userId: userId, idToken: idToken, rawToken: token)
    }

    public func tokenRefreshed(_ token: String) async throws {
        guard let previous = registration else { return }
        try await register(userId: previous.userId, idToken: previous.idToken, rawToken: token)
        let newId = FcmToken.id(for: token)
        if previous.tokenId != newId {
            try await store.delete(
                userId: previous.userId,
                tokenId: previous.tokenId,
                idToken: previous.idToken
            )
        }
    }

    private func register(userId: String, idToken: String, rawToken: String) async throws {
        guard !rawToken.isEmpty else { return }
        let token = FcmToken(
            id: FcmToken.id(for: rawToken),
            userId: userId,
            token: rawToken,
            os: .iOS,
            environment: environment,
            updatedAt: now()
        )
        try await store.save(token, idToken: idToken)
        registration = Registration(userId: userId, idToken: idToken, tokenId: token.id)
    }
}

public struct FirestoreFcmTokenStore: FcmTokenStoring {
    private let projectId: String
    private let session: URLSession

    public init(projectId: String, session: URLSession = .shared) {
        self.projectId = projectId
        self.session = session
    }

    public func save(_ token: FcmToken, idToken: String) async throws {
        let fields: [String: Any] = [
            "id": stringValue(token.id),
            "userId": stringValue(token.userId),
            "token": stringValue(token.token),
            "appVariant": stringValue(token.appVariant),
            "os": stringValue(token.os.rawValue),
            "environment": stringValue(token.environment.rawValue),
            "updatedAt": ["timestampValue": ISO8601DateFormatter().string(from: token.updatedAt)]
        ]
        try await request(
            userId: token.userId,
            tokenId: token.id,
            method: "PATCH",
            idToken: idToken,
            body: ["fields": fields]
        )
    }

    public func delete(userId: String, tokenId: String, idToken: String) async throws {
        try await request(
            userId: userId,
            tokenId: tokenId,
            method: "DELETE",
            idToken: idToken,
            body: nil
        )
    }

    private func request(
        userId: String,
        tokenId: String,
        method: String,
        idToken: String,
        body: [String: Any]?
    ) async throws {
        let escapedUser = userId.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? userId
        let escapedToken = tokenId.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? tokenId
        let value = "https://firestore.googleapis.com/v1/projects/\(projectId)"
            + "/databases/(default)/documents/memberPrivate/\(escapedUser)/fcmTokens/\(escapedToken)"
        var request = URLRequest(url: URL(string: value)!)
        request.httpMethod = method
        request.setValue("Bearer \(idToken)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let body { request.httpBody = try JSONSerialization.data(withJSONObject: body) }
        let (_, response) = try await session.data(for: request)
        let status = (response as? HTTPURLResponse)?.statusCode ?? 0
        guard (200..<300).contains(status) || (method == "DELETE" && status == 404) else {
            throw URLError(.badServerResponse)
        }
    }

    private func stringValue(_ value: String) -> [String: String] {
        ["stringValue": value]
    }
}

public extension Notification.Name {
    static let fcmTokenRefreshed = Notification.Name("OrgPortalNext.fcmTokenRefreshed")
}

private extension String {
    var nilIfEmpty: String? { isEmpty ? nil : self }
}

File: NextApp/Android/core/notifications/src/main/kotlin/jp/komehyappyo/member/next/core/notifications/NotificationService.kt
package jp.komehyappyo.member.next.core.notifications

import android.Manifest
import android.app.AlarmManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import android.net.Uri
import androidx.core.content.ContextCompat
import com.google.firebase.FirebaseApp
import com.google.firebase.FirebaseOptions
import com.google.firebase.messaging.FirebaseMessaging
import com.google.firebase.messaging.FirebaseMessagingService
import com.google.firebase.messaging.RemoteMessage
import jp.komehyappyo.member.next.core.model.Schedule
import java.net.HttpURLConnection
import java.net.URI
import java.security.MessageDigest
import java.time.Instant
import java.util.concurrent.CopyOnWriteArraySet
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withContext
import org.json.JSONObject
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

data class FcmToken(
    val id: String,
    val userId: String,
    val token: String,
    val appVariant: String = "next",
    val os: String = "Android",
    val environment: String,
    val updatedAt: Instant,
) {
    companion object {
        fun idFor(token: String): String = MessageDigest.getInstance("SHA-256")
            .digest(token.toByteArray(Charsets.UTF_8))
            .joinToString("") { "%02x".format(it.toInt() and 0xff) }
    }
}

enum class NotificationType(val payloadValue: String, val host: String) {
    Announcement("announcement", "announcements"),
    AdminReply("admin_reply", "posts"),
    VideoQuestionAnswer("video_question_answer", "video-questions"),
    Event("event", "events"),
    SupportMessage("support_message", "support"),
}

data class NotificationRoute(
    val type: NotificationType,
    val targetId: String,
    val communityId: String? = null,
)

data class NotificationNavigationDecision(
    val route: NotificationRoute,
    val communityIdToSelect: String?,
) {
    companion object {
        fun resolve(route: NotificationRoute, selectedCommunityId: String?): NotificationNavigationDecision =
            NotificationNavigationDecision(
                route = route,
                communityIdToSelect = route.communityId?.takeIf { it != selectedCommunityId },
            )
    }
}

object NotificationRouteParser {
    const val TYPE_EXTRA = "notificationType"
    const val DEEP_LINK_EXTRA = "deepLink"
    const val COMMUNITY_ID_EXTRA = "communityId"
    private val targetKeys = mapOf(
        NotificationType.Announcement to "announcementId",
        NotificationType.AdminReply to "postId",
        NotificationType.VideoQuestionAnswer to "questionId",
        NotificationType.Event to "eventId",
        NotificationType.SupportMessage to "supportMessageId",
    )

    fun uri(route: NotificationRoute): Uri {
        val path = if (route.type == NotificationType.SupportMessage) {
            "messages/${route.targetId}"
        } else {
            route.targetId
        }
        return Uri.Builder()
            .scheme("orgportalnext")
            .authority(route.type.host)
            .path(path)
            .apply { route.communityId?.let { appendQueryParameter(COMMUNITY_ID_EXTRA, it) } }
            .build()
    }

    fun route(intent: Intent?): NotificationRoute? {
        if (intent == null) return null
        intent.getStringExtra(DEEP_LINK_EXTRA)?.let { value ->
            route(Uri.parse(value))?.let { return it }
        }
        route(intent.data)?.let { return it }
        val type = NotificationType.entries.firstOrNull {
            it.payloadValue == intent.getStringExtra(TYPE_EXTRA)
        } ?: return null
        val targetId = intent.getStringExtra(targetKeys.getValue(type))
            ?: intent.getStringExtra("targetId")
        return targetId?.takeIf(String::isNotBlank)?.let {
            NotificationRoute(type, it, intent.getStringExtra(COMMUNITY_ID_EXTRA)?.takeIf(String::isNotBlank))
        }
    }

    fun route(uri: Uri?): NotificationRoute? {
        if (uri?.scheme != "orgportalnext") return null
        val type = NotificationType.entries.firstOrNull { it.host == uri.host } ?: return null
        val segments = uri.pathSegments
        val targetId = when (type) {
            NotificationType.SupportMessage -> segments.takeIf {
                it.size == 2 && it[0] == "messages"
            }?.get(1)
            else -> segments.singleOrNull()
        }?.takeIf(String::isNotBlank) ?: return null
        return NotificationRoute(
            type = type,
            targetId = targetId,
            communityId = uri.getQueryParameter(COMMUNITY_ID_EXTRA)?.takeIf(String::isNotBlank),
        )
    }

    fun putExtras(intent: Intent, route: NotificationRoute): Intent = intent.apply {
        putExtra(TYPE_EXTRA, route.type.payloadValue)
        putExtra(targetKeys.getValue(route.type), route.targetId)
        route.communityId?.let { putExtra(COMMUNITY_ID_EXTRA, it) }
        data = uri(route)
    }
}

object SupportNotificationRoute {
    const val MESSAGE_ID_EXTRA = "supportMessageId"
    const val TYPE_EXTRA = NotificationRouteParser.TYPE_EXTRA
    const val TYPE_VALUE = "support_message"

    fun uri(messageId: String): Uri = NotificationRouteParser.uri(
        NotificationRoute(NotificationType.SupportMessage, messageId),
    )

    fun messageId(intent: Intent?): String? {
        intent?.getStringExtra(MESSAGE_ID_EXTRA)?.takeIf(String::isNotBlank)?.let { return it }
        return NotificationRouteParser.route(intent)
            ?.takeIf { it.type == NotificationType.SupportMessage }
            ?.targetId
    }
}

interface FcmTokenStore {
    suspend fun save(token: FcmToken, idToken: String)
    suspend fun delete(userId: String, tokenId: String, idToken: String)
}

fun interface FcmRegistrationTokenProvider {
    suspend fun currentToken(): String
}

object FirebaseMessagingRuntime {
    fun configure(
        context: Context,
        projectId: String,
        apiKey: String,
        applicationId: String,
        senderId: String,
    ): Boolean {
        if (FirebaseApp.getApps(context).isNotEmpty()) return true
        if (applicationId.isBlank() || senderId.isBlank()) return false
        val options = FirebaseOptions.Builder()
            .setProjectId(projectId)
            .setApiKey(apiKey)
            .setApplicationId(applicationId)
            .setGcmSenderId(senderId)
            .build()
        FirebaseApp.initializeApp(context, options)
        return true
    }

    fun setAutoInitEnabled(enabled: Boolean) {
        FirebaseMessaging.getInstance().isAutoInitEnabled = enabled
    }
}

class FirebaseRegistrationTokenProvider : FcmRegistrationTokenProvider {
    override suspend fun currentToken(): String = suspendCancellableCoroutine { continuation ->
        FirebaseMessaging.getInstance().token
            .addOnSuccessListener { continuation.resume(it) }
            .addOnFailureListener { continuation.resumeWithException(it) }
    }
}

class FcmTokenRegistrationManager(
    private val store: FcmTokenStore,
    private val tokenProvider: FcmRegistrationTokenProvider,
    private val environment: String,
    private val now: () -> Instant = Instant::now,
) {
    private data class Registration(val userId: String, val idToken: String, val tokenId: String)
    private var registration: Registration? = null

    suspend fun synchronize(userId: String?, idToken: String?) {
        if (userId.isNullOrBlank() || userId == "guest" || idToken.isNullOrBlank()) {
            registration?.let { store.delete(it.userId, it.tokenId, it.idToken) }
            registration = null
            return
        }
        registration?.takeIf { it.userId != userId }?.let {
            store.delete(it.userId, it.tokenId, it.idToken)
            registration = null
        }
        register(userId, idToken, tokenProvider.currentToken())
    }

    suspend fun tokenRefreshed(token: String) {
        val current = registration ?: return
        register(current.userId, current.idToken, token)
        if (current.tokenId != FcmToken.idFor(token)) {
            store.delete(current.userId, current.tokenId, current.idToken)
        }
    }

    private suspend fun register(userId: String, idToken: String, rawToken: String) {
        if (rawToken.isBlank()) return
        val token = FcmToken(
            id = FcmToken.idFor(rawToken),
            userId = userId,
            token = rawToken,
            environment = environment,
            updatedAt = now(),
        )
        store.save(token, idToken)
        registration = Registration(userId, idToken, token.id)
    }
}

class FirestoreFcmTokenStore(private val projectId: String) : FcmTokenStore {
    override suspend fun save(token: FcmToken, idToken: String) {
        request(token.userId, token.id, "PATCH", idToken, token.firestoreBody())
    }

    override suspend fun delete(userId: String, tokenId: String, idToken: String) {
        request(userId, tokenId, "DELETE", idToken, null)
    }

    private suspend fun request(
        userId: String,
        tokenId: String,
        method: String,
        idToken: String,
        body: JSONObject?,
    ) = withContext(Dispatchers.IO) {
        val encodedUser = Uri.encode(userId)
        val encodedToken = Uri.encode(tokenId)
        val url = URI(
            "https://firestore.googleapis.com/v1/projects/$projectId/databases/(default)/documents/" +
                "memberPrivate/$encodedUser/fcmTokens/$encodedToken",
        ).toURL()
        val connection = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = method
            setRequestProperty("Authorization", "Bearer $idToken")
            setRequestProperty("Content-Type", "application/json")
            connectTimeout = 15_000
            readTimeout = 15_000
            if (body != null) {
                doOutput = true
                outputStream.use { it.write(body.toString().toByteArray(Charsets.UTF_8)) }
            }
        }
        val status = connection.responseCode
        if (status !in 200..299 && !(method == "DELETE" && status == 404)) {
            val message = connection.errorStream?.bufferedReader()?.use { it.readText() }
            error("FCM token Firestore request failed ($status): ${message.orEmpty()}")
        }
        connection.disconnect()
    }

    private fun FcmToken.firestoreBody(): JSONObject = JSONObject().put(
        "fields",
        JSONObject()
            .put("id", stringValue(id))
            .put("userId", stringValue(userId))
            .put("token", stringValue(token))
            .put("appVariant", stringValue(appVariant))
            .put("os", stringValue(os))
            .put("environment", stringValue(environment))
            .put("updatedAt", JSONObject().put("timestampValue", updatedAt.toString())),
    )

    private fun stringValue(value: String) = JSONObject().put("stringValue", value)
}

object FcmTokenRefreshBus {
    private val listeners = CopyOnWriteArraySet<(String) -> Unit>()
    fun add(listener: (String) -> Unit) { listeners += listener }
    fun remove(listener: (String) -> Unit) { listeners -= listener }
    fun publish(token: String) { listeners.forEach { it(token) } }
}

class NextFirebaseMessagingService : FirebaseMessagingService() {
    override fun onNewToken(token: String) {
        super.onNewToken(token)
        FcmTokenRefreshBus.publish(token)
    }

    override fun onMessageReceived(message: RemoteMessage) {
        val route = route(message.data) ?: return
        NotificationService(this).showRemoteNotification(
            route = route,
            title = message.notification?.title ?: "次期会員アプリ",
            body = message.notification?.body ?: "新しいお知らせがあります。",
        )
    }

    private fun route(data: Map<String, String>): NotificationRoute? {
        val intent = Intent()
        data.forEach { (key, value) -> intent.putExtra(key, value) }
        return NotificationRouteParser.route(intent)
    }
}

class NotificationService(private val context: Context) {
    fun hasPermission(): Boolean =
        Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU ||
            ContextCompat.checkSelfPermission(
                context,
                Manifest.permission.POST_NOTIFICATIONS,
            ) == PackageManager.PERMISSION_GRANTED

    fun scheduleReminder(schedule: Schedule) {
        val reminder = schedule.reminderSetting?.takeIf { it.isEnabled } ?: return
        val triggerAt = schedule.startDateTime
            .minusSeconds(reminder.notifyBeforeMinutes * 60L)
            .toEpochMilli()
        if (triggerAt <= System.currentTimeMillis()) return

        val intent = Intent(context, ScheduleReminderReceiver::class.java)
            .putExtra("title", schedule.title)
        val pendingIntent = PendingIntent.getBroadcast(
            context,
            schedule.id.hashCode(),
            intent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE,
        )
        val alarmManager = context.getSystemService(AlarmManager::class.java)
        alarmManager.setAndAllowWhileIdle(AlarmManager.RTC_WAKEUP, triggerAt, pendingIntent)
    }

    @android.annotation.SuppressLint("MissingPermission")
    fun showRemoteNotification(route: NotificationRoute, title: String, body: String) {
        if (!hasPermission()) return
        val launchIntent = context.packageManager.getLaunchIntentForPackage(context.packageName)
            ?: return
        NotificationRouteParser.putExtras(launchIntent, route)
        val pendingIntent = PendingIntent.getActivity(
            context,
            route.hashCode(),
            launchIntent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE,
        )
        val manager = context.getSystemService(android.app.NotificationManager::class.java)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            manager.createNotificationChannel(
                android.app.NotificationChannel(
                    REMOTE_CHANNEL_ID,
                    "お知らせ",
                    android.app.NotificationManager.IMPORTANCE_DEFAULT,
                ),
            )
        }
        val notification = androidx.core.app.NotificationCompat.Builder(context, REMOTE_CHANNEL_ID)
            .setSmallIcon(android.R.drawable.ic_dialog_info)
            .setContentTitle(title)
            .setContentText(body)
            .setAutoCancel(true)
            .setContentIntent(pendingIntent)
            .build()
        androidx.core.app.NotificationManagerCompat.from(context)
            .notify(route.hashCode(), notification)
    }

    fun showSupportMessageNotification(messageId: String) {
        showRemoteNotification(
            route = NotificationRoute(NotificationType.SupportMessage, messageId),
            title = "サポート",
            body = "サポートから新しいメッセージがあります",
        )
    }

    private companion object {
        const val REMOTE_CHANNEL_ID = "remote_notifications"
    }
}
